### PR TITLE
Fix lingering PolyMC icon in Dock on macOS

### PR DESCRIPTION
**Pull request by ryanccn**
This PR ~~makes the app on macOS only quit when manually told by the user to do so~~.

It also fixes the weird behavior when you launch an instance, close the main window, but the PolyMC icon stays in the dock and is not responsive.

### CODE REVIEW

**Comment by kthchew** [comment]
I don't think there's actually really a standard for this (speaking about the only quitting when manually doing so). The best pattern I can see among Apple's apps is that apps with a main window (e.g. not a document-based app that can have multiple documents open) seem to quit when their windows are closed. See, for example, System Preferences, Reminders, and Disk Utility. But this falls apart with some other apps, such as Activity Monitor (actually, I guess there's an excuse there since you can have a dynamic dock icon, but still). 

Anyway, IMO since everything revolves around the main window with the instances, I'd argue that the current behavior could actually be more "standard." (You wouldn't open a bunch of instance settings like you open documents in Word, and instance settings are basically "children" of the main window anyway.) But the standard seems so inconsistent that I feel like it's not worth making this change, unless there is some argument for PolyMC being able to run in the background (e.g. if automatic mod updates come around and people want it to do it in the background. Maybe not the greatest example but you get the idea).

The bug fix is good though.

**Comment by ryanccn** [comment]
Hmm, but the bugfix is associated with this dissociation of `MainWindow` and `Application`. So I would argue adopting the "(not) standard" would be an overall improvement.

~~Also, can instances count as documents~~

**Comment by ryanccn** [comment]
Discord actually just hides itself when you "close" the window... very non-standard

**Comment by kthchew** [comment]
> Also, can instances count as documents

Well, instances look (and basically are) separate apps on the Dock, so maybe not 😛 

> Discord actually just hides itself when you "close" the window... very non-standard

Yes, the "standard" is not really followed (even by Apple). It's so bad that I can't even tell what the standard is supposed to be lol

**Comment by ryanccn** [comment]
fwiw on Electron apps are generally expected to do this:

```typescript
app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

app.on('activate', () => {
  if (BrowserWindow.getAllWindows().length === 0) createWindow();
});
```

**Comment by ryanccn** [comment]
I mostly created this PR because the dock icon remained in Dock even though PolyMC has stopped responding to anything after the main window has been closed, so that's still pretty valid.

About the quitting thing, it's pretty optional. If we decide not to implement this behavior, just removing these lines should work.

https://github.com/ryanccn/PolyMC/blob/9673dac22b0ff81a54847d5db5438c099a6df587/launcher/Application.cpp#L1309-L1311

**Comment by ryanccn** [comment]
no one cares about macOS 😶‍🌫️

**Review by flowln — APPROVED** [review]

File: launcher/Application.cpp [hunk 1]
```diff
@@ -954,6 +960,22 @@ bool Application::createSetupWizard()
...
 }
 
+bool Application::event(QEvent* event) {
+#ifdef Q_OS_MACOS
+    if (event->type() == QEvent::ApplicationStateChange) {
+        auto ev = static_cast<QApplicationStateChangeEvent*>(event);
+
+        if (m_prevAppState == Qt::ApplicationActive
+                && ev->applicationState() == Qt::ApplicationActive) {
+            qDebug() << "Clicked on dock!";
```
Comment (Scrumplex) [inline launcher/Application.cpp]: this should probably go

**Review by d-513 — APPROVED** [review]

**Review by timoreo22 — APPROVED** [review]

**Review by Scrumplex — APPROVED** [review]

**Pull request merged**